espaloma/graphs/deploy.py

```diff
@@ -0,0 +1,171 @@
+                    ).numpy().item()
+
+                _eq = Quantity(
+                    _eq,
+                    esp.units.ANGLE_UNIT,
+                ).value_in_unit(OPENMM_ANGLE_EQ_UNIT)
+
+                _k = 2.0 * Quantity(
+                    _k,
+                    esp.units.ANGLE_FORCE_CONSTANCE_UNIT,
+                ).value_in_unit(OPENMM_ANGLE_K_UNIT)
+
+                force.setAngleParameters(idx, idx0, idx1, idx2, _eq, _k)
+
+        if 'PeriodicTorsionForce' in name:
+            number_of_torsions = force.getNumTorsions()
+            blank_periodicity = False
+            if 'periodicity%s' % suffix not in g.nodes['n4'].data\
+                or 'phase%s' % suffix not in g.nodes['n4'].data:
+                blank_periodicity = True
+
+                g.nodes['n4'].data['periodicity%s' % suffix] = torch.arange(
+                    1, 6
+                )[None, :].repeat(
+                    g.heterograph.number_of_nodes('n4'), 1
+                )
+
+                g.nodes['n4'].data['phases%s' % suffix] = torch.zeros(
+                    g.heterograph.number_of_nodes('n4'),
+                    6
+                )
+
+            count_idx = 0
+            for idx in range(g.heterograph.number_of_nodes('n4')):
+                idx0 = g.nodes['n4'].data['idxs'][idx, 0].item()
+                idx1 = g.nodes['n4'].data['idxs'][idx, 1].item()
+                idx2 = g.nodes['n4'].data['idxs'][idx, 2].item()
+                idx3 = g.nodes['n4'].data['idxs'][idx, 3].item()
+
+                if idx0 < idx3:
+                    periodicities = g.nodes['n4'].data['periodicity%s' % suffix][idx]
+                    phases = g.nodes['n4'].data['phases%s' % suffix][idx]
+                    ks = g.nodes['n4'].data['k%s' % suffix][idx]
+                    for sub_idx in range(ks.flatten().shape[0]):
+                        k = ks[sub_idx].item()
+                        if k != 0.0:
+                            _periodicity = periodicities[sub_idx].item()
+                            _phase = phases[sub_idx].item()
+
+                            k = Quantity(
+                                k,
+                                esp.units.ENERGY_UNIT,
+                            ).value_in_unit(
+                                OPENMM_ENERGY_UNIT,
+                            )
+
+                            if count_idx < number_of_torsions:
+                                force.setTorsionParameters(
+                                    count_idx,
+                                    idx0, idx1, idx2, idx3,
+                                    _periodicity, _phase, 2.0 * k,
+                                )
+
+                            else:
+                                force.addTorsion(
+                                    idx0, idx1, idx2, idx3,
+                                    _periodicity, _phase, 2.0 * k,
+                                )
+
+                            count_idx += 1
```

Comment:
These cover the conditions where `number_of_torsions` <= `g.heterograph.number_of_nodes('n4')`.

Possible missing condition: what happens if `number_of_torsions` > `g.heterograph.number_of_nodes('n4')`?

May want to start with an empty PeriodicTorsionForce and always call `force.addTorsion`, or go afterwards and delete / zero out any torsions with index >= `g.heterograph.number_of_nodes('n4')`.

Reply:
How can it be greater

Reply:
I made those awkward logic because OpenMM doesn't allow deleting torsions.

Reply:
> How can it be greater

Not sure if it ever is for our molecules and this forcefield, but it could in principle be greater, since we encode periodicities 1-6 in a single vector sitting on each torsion, but PeriodicTorsionForce would need up to 6 terms per torsion to do the same.

Reply:
>I made those awkward logic because OpenMM doesn't allow deleting torsions.

An alternative would be to start with an empty PeriodicTorsionForce and always call `force.addTorsion`

Reply:
Resolved by asserting `number_of_torsions <= g.heterograph.number_of_nodes('n4')` in https://github.com/choderalab/espaloma/pull/39/commits/8cddf41708713a215f6d9baf1a4747f1506ebad2